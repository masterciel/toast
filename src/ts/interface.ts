import _Vue from "vue";
import ToastContainer from "../components/VtToastContainer.vue";
import events from "./events";
import { TYPE, EVENTS } from "./constants";
import { getId } from "./utils";
import { ToastContent, ToastOptions, ToastID, PluginOptions } from "../types";

const ToastInterface = (Vue: typeof _Vue, globalOptions?: PluginOptions) => {
  new (Vue.extend(ToastContainer))({
    el: document.createElement("div"),
    propsData: globalOptions
  });
  /**
   * Display a toast
   */
  const toast = (content: ToastContent, options?: ToastOptions): ToastID => {
    const props = Object.assign(
      {},
      { id: getId(), type: TYPE.DEFAULT },
      options,
      { content }
    );
    events.$emit(EVENTS.ADD, props);
    return props.id;
  };
  /**
   * Clear all toasts
   */
  toast.clear = () => events.$emit(EVENTS.CLEAR);
  /**
   * Update Plugin Defaults
   */
  toast.updateDefaults = (update: PluginOptions) => {
    events.$emit(EVENTS.UPDATE_DEFAULTS, update);
  };
  /**
   * Dismiss toast specified by an id
   */
  toast.dismiss = (id: ToastID) => {
    events.$emit(EVENTS.DISMISS, id);
  };
  /**
   * Update Toast
   */
  toast.update = (
    id: ToastID,
    { content, options }: { content?: ToastContent; options?: ToastOptions },
    create = false
  ) => {
    events.$emit(EVENTS.UPDATE, {
      id,
      options: Object.assign({}, options, { content }),
      create
    });
  };
  /**
   * Display a success toast
   */
  toast.success = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.SUCCESS }
  ) => toast(content, Object.assign({}, options, { type: TYPE.SUCCESS }));

  /**
   * Display an info toast
   */
  toast.info = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.INFO }
  ) => toast(content, Object.assign({}, options, { type: TYPE.INFO }));

  /**
   * Display an error toast
   */
  toast.error = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.ERROR }
  ) => toast(content, Object.assign({}, options, { type: TYPE.ERROR }));

  /**
   * Display a warning toast
   */
  toast.warning = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.WARNING }
  ) => toast(content, Object.assign({}, options, { type: TYPE.WARNING }));

  return toast;
};

export default ToastInterface;
