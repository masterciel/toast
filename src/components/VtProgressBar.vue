<template>
  <div ref="el" :style="style" :class="cpClass" />
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  ref,
  watch,
  nextTick,
  onMounted,
  onBeforeUnmount,
} from "vue"

import { VT_NAMESPACE } from "../ts/constants"
import PROPS from "../ts/propValidators"

export default defineComponent({
  name: "VtProgressBar",

  props: PROPS.PROGRESS_BAR,

  emits: ["close-toast"],

  setup(props, { emit }) {
    const el = ref<HTMLElement>()
    const hasClass = ref(true)

    const style = computed(() => {
      return {
        animationDuration: `${props.timeout}ms`,
        animationPlayState: props.isRunning ? "running" : "paused",
        opacity: props.hideProgressBar ? 0 : 1,
      }
    })

    const cpClass = computed(() =>
      hasClass.value ? `${VT_NAMESPACE}__progress-bar` : ""
    )

    watch(
      () => props.timeout,
      () => {
        hasClass.value = false
        nextTick(() => (hasClass.value = true))
      }
    )

    const animationEnded = () => emit("close-toast")

    onMounted(() => {
      /* istanbul ignore else  */
      if (el.value) {
        el.value.addEventListener("animationend", animationEnded)
      }
    })
    onBeforeUnmount(() => {
      /* istanbul ignore else  */
      if (el.value) {
        el.value.removeEventListener("animationend", animationEnded)
      }
    })

    return {
      el,
      style,
      cpClass,
    }
  },
})
</script>
