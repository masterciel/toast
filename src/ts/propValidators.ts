import { PropType } from "vue";
import {
  ToastContent,
  CommonOptions,
  PluginOptions,
  ToastID,
  ToastOptions
} from "../types";
import { TYPE, POSITION, VT_NAMESPACE } from "./constants";

const COMMON = {
  type: {
    type: String as PropType<TYPE>,
    default: TYPE.DEFAULT
  },
  classNames: {
    type: [String, Array] as PropType<string | string[]>,
    default: () => []
  },
  trueBoolean: {
    type: Boolean,
    default: true
  }
};

const ICON = {
  type: COMMON.type,
  customIcon: {
    type: [String, Boolean, Object, Function] as PropType<
      NonNullable<CommonOptions["icon"]>
    >,
    default: true
  }
};

const CLOSE_BUTTON = {
  component: {
    type: [String, Object, Function, Boolean] as PropType<
      NonNullable<CommonOptions["closeButton"]>
    >,
    default: "button"
  },
  classNames: COMMON.classNames,
  showOnHover: Boolean,
  ariaLabel: {
    type: String,
    default: "close"
  }
};

const PROGRESS_BAR = {
  timeout: {
    type: [Number, Boolean] as PropType<number | false>,
    default: 5000
  },
  hideProgressBar: Boolean,
  isRunning: Boolean
};

const TRANSITION = {
  transition: {
    type: [Object, String] as PropType<
      NonNullable<PluginOptions["transition"]>
    >,
    default: `${VT_NAMESPACE}__bounce`
  },
  transitionDuration: {
    type: [Number, Object] as PropType<
      NonNullable<PluginOptions["transitionDuration"]>
    >,
    default: 750
  }
};

const CORE_TOAST = {
  position: {
    type: String as PropType<POSITION>,
    default: POSITION.TOP_RIGHT
  },
  draggable: COMMON.trueBoolean,
  draggablePercent: {
    type: Number,
    default: 0.6
  },
  pauseOnFocusLoss: COMMON.trueBoolean,
  pauseOnHover: COMMON.trueBoolean,
  closeOnClick: COMMON.trueBoolean,
  timeout: PROGRESS_BAR.timeout,
  hideProgressBar: PROGRESS_BAR.hideProgressBar,
  toastClassName: COMMON.classNames,
  bodyClassName: COMMON.classNames,
  icon: ICON.customIcon,
  closeButton: CLOSE_BUTTON.component,
  closeButtonClassName: CLOSE_BUTTON.classNames,
  showCloseButtonOnHover: CLOSE_BUTTON.showOnHover,
  accessibility: {
    type: Object as PropType<PluginOptions["accessibility"]>,
    default: () => ({
      toastRole: "alert",
      closeButtonLabel: "close"
    })
  },
  rtl: Boolean
};

const TOAST = {
  id: {
    type: [String, Number] as PropType<ToastID>,
    required: true
  },
  type: COMMON.type,
  content: {
    type: [String, Object, Function] as PropType<ToastContent>,
    required: true
  },
  onClick: Function as PropType<NonNullable<ToastOptions["onClick"]>>,
  onClose: Function as PropType<NonNullable<ToastOptions["onClose"]>>
};

const CONTAINER = {
  newestOnTop: COMMON.trueBoolean,
  maxToasts: {
    type: Number,
    default: 20
  },
  container: {
    type: (Element as unknown) as PropType<
      NonNullable<PluginOptions["container"]>
    >,
    default: () => document.body
  },
  filterBeforeCreate: Function as PropType<
    NonNullable<PluginOptions["filterBeforeCreate"]>
  >,
  filterToasts: Function as PropType<
    NonNullable<PluginOptions["filterToasts"]>
  >,
  containerClassName: COMMON.classNames
};

export default {
  CORE_TOAST,
  TOAST,
  CONTAINER,
  PROGRESS_BAR,
  ICON,
  TRANSITION,
  CLOSE_BUTTON
};
