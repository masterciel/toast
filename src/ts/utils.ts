import { Component, defineComponent } from "vue"
import type {
  ToastComponent,
  ToastContent,
  RenderableToastContent,
} from "../types"

interface DictionaryLike {
  [index: string]: unknown
}

// eslint-disable-next-line @typescript-eslint/ban-types
const isFunction = (value: unknown): value is Function =>
  typeof value === "function"

const isString = (value: unknown): value is string => typeof value === "string"

const isNonEmptyString = (value: unknown): value is string =>
  isString(value) && value.trim().length > 0

const isNumber = (value: unknown): value is number => typeof value === "number"

const isUndefined = (value: unknown): value is undefined =>
  typeof value === "undefined"

const isObject = (value: unknown): value is DictionaryLike =>
  typeof value === "object" && value !== null

const isJSX = (obj: unknown): obj is JSX.Element =>
  hasProp(obj, "tag") && isNonEmptyString(obj.tag)

const isTouchEvent = (event: Event): event is TouchEvent =>
  window.TouchEvent && event instanceof TouchEvent

const isToastComponent = (obj: unknown): obj is ToastComponent =>
  hasProp(obj, "component") && isToastContent(obj.component)

const isVueComponent = (c: unknown): c is Component => {
  if (isFunction(c)) {
    return true
  }
  return isObject(c)
}

const isToastContent = (obj: unknown): obj is ToastContent =>
  // Ignore undefined
  !isUndefined(obj) &&
  // Is a string
  (isString(obj) ||
    // Regular Vue component
    isVueComponent(obj) ||
    // Nested object
    isToastComponent(obj))

const isDOMRect = (obj: unknown): obj is DOMRect =>
  isObject(obj) &&
  ["height", "width", "right", "left", "top", "bottom"].every(p =>
    isNumber(obj[p])
  )

const hasProp = <O extends unknown, K extends PropertyKey>(
  obj: O,
  propKey: K
): obj is O & { [key in K]: unknown } =>
  Object.prototype.hasOwnProperty.call(obj, propKey)

/**
 * ID generator
 */
const getId = (i => () => i++)(0)

function getX(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientX : event.clientX
}

function getY(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientY : event.clientY
}

const removeElement = (el: Element) => {
  if (!isUndefined(el.remove)) {
    el.remove()
  } else if (el.parentNode) {
    el.parentNode.removeChild(el)
  }
}

const getVueComponentFromObj = (obj: ToastContent): RenderableToastContent => {
  if (isToastComponent(obj)) {
    // Recurse if component prop
    return getVueComponentFromObj(obj.component)
  }
  if (isJSX(obj)) {
    // Create render function for JSX
    return defineComponent({
      render() {
        return obj
      },
    })
  }
  // Return the actual object if regular vue component
  return obj
}

export {
  getId,
  getX,
  getY,
  removeElement,
  isString,
  isNonEmptyString,
  isToastContent,
  getVueComponentFromObj,
  hasProp,
  isUndefined,
  isDOMRect,
  isFunction,
}
