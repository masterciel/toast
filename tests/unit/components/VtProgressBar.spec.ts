import { mount } from "@vue/test-utils";
import VtProgressBar from "../../../src/components/VtProgressBar.vue";
import { VT_NAMESPACE } from "../../../src/ts/constants";

describe("VtProgressBar", () => {
  it("matches snapshot", () => {
    const wrapper = mount(VtProgressBar);
    expect(wrapper.element).toMatchSnapshot();
  });
  it("has default class", () => {
    const wrapper = mount(VtProgressBar);
    expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__progress-bar`);
  });
  it("has default style values", () => {
    const wrapper = mount(VtProgressBar);
    const vm = wrapper.vm as { [index: string]: unknown };
    const style = vm.style as {
      animationDuration: string;
      animationPlayState: string;
      opacity: 0 | 1;
    };
    expect(style).toEqual({
      animationDuration: "5000ms",
      animationPlayState: "paused",
      opacity: 1
    });
  });
  it("sets style duration from timeout", () => {
    const wrapper = mount(VtProgressBar, {
      propsData: {
        timeout: 1000
      }
    });
    const vm = wrapper.vm as { [index: string]: unknown };
    const style = vm.style as {
      animationDuration: string;
      animationPlayState: string;
      opacity: 0 | 1;
    };
    expect(style).toEqual({
      animationDuration: "1000ms",
      animationPlayState: "paused",
      opacity: 1
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("sets playstate from isRunning", () => {
    const wrapper = mount(VtProgressBar, {
      propsData: {
        isRunning: true
      }
    });
    const vm = wrapper.vm as { [index: string]: unknown };
    const style = vm.style as {
      animationDuration: string;
      animationPlayState: string;
      opacity: 0 | 1;
    };
    expect(style).toEqual({
      animationDuration: "5000ms",
      animationPlayState: "running",
      opacity: 1
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("sets opacity to 0 from from hideProgressBar", () => {
    const wrapper = mount(VtProgressBar, {
      propsData: {
        hideProgressBar: true
      }
    });
    const vm = wrapper.vm as { [index: string]: unknown };
    const style = vm.style as {
      animationDuration: string;
      animationPlayState: string;
      opacity: 0 | 1;
    };
    expect(style).toEqual({
      animationDuration: "5000ms",
      animationPlayState: "paused",
      opacity: 0
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("triggers class reset on timeout change", async () => {
    const wrapper = mount(VtProgressBar);
    expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__progress-bar`);
    wrapper.setProps({ timeout: 1000 });
    await wrapper.vm.$nextTick();
    expect(wrapper.classes()).not.toContain(`${VT_NAMESPACE}__progress-bar`);
    expect(wrapper.element).toMatchSnapshot();
    await wrapper.vm.$nextTick();
    expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__progress-bar`);
    expect(wrapper.element).toMatchSnapshot();
  });
  it("emits close-toast on animationend", async () => {
    const wrapper = mount(VtProgressBar);
    expect(wrapper.emitted("close-toast")).toBeFalsy();
    wrapper.trigger("animationend");
    expect(wrapper.emitted("close-toast")).toBeTruthy();
  });
  it("removes listener on beforeDestroy", async () => {
    const wrapper = mount(VtProgressBar);
    const spyRemoveEventListener = jest.spyOn(
      wrapper.vm.$el,
      "removeEventListener"
    );
    expect(spyRemoveEventListener).not.toHaveBeenCalled();
    wrapper.destroy();
    expect(spyRemoveEventListener).toHaveBeenCalled();
  });
});
