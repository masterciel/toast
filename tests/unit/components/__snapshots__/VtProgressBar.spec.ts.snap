// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VtProgressBar matches snapshot 1`] = `
<div
  class="Vue-Toastification__progress-bar"
  style="animation-duration: 5000ms; animation-play-state: paused; opacity: 1;"
/>
`;

exports[`VtProgressBar sets opacity to 0 from from hideProgressBar 1`] = `
<div
  class="Vue-Toastification__progress-bar"
  style="animation-duration: 5000ms; animation-play-state: paused; opacity: 0;"
/>
`;

exports[`VtProgressBar sets playstate from isRunning 1`] = `
<div
  class="Vue-Toastification__progress-bar"
  style="animation-duration: 5000ms; animation-play-state: running; opacity: 1;"
/>
`;

exports[`VtProgressBar sets style duration from timeout 1`] = `
<div
  class="Vue-Toastification__progress-bar"
  style="animation-duration: 1000ms; animation-play-state: paused; opacity: 1;"
/>
`;

exports[`VtProgressBar triggers class reset on timeout change 1`] = `
<div
  class=""
  style="animation-duration: 1000ms; animation-play-state: paused; opacity: 1;"
/>
`;

exports[`VtProgressBar triggers class reset on timeout change 2`] = `
<div
  class="Vue-Toastification__progress-bar"
  style="animation-duration: 1000ms; animation-play-state: paused; opacity: 1;"
/>
`;
